tests/integration/main_test.go

```diff
@@ -0,0 +1,592 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const testNamespace = "foo"
+
+type IntegrationSuite struct {
+	suite.Suite
+	clientFactory client.SharedClientFactory
+}
+
+func TestIntegrationTest(t *testing.T) {
+	suite.Run(t, new(IntegrationSuite))
+}
+
+func (m *IntegrationSuite) SetupSuite() {
+	logrus.SetLevel(logrus.DebugLevel)
+	kubeconfigPath := os.Getenv("KUBECONFIG")
+	logrus.Infof("Setting up test with KUBECONFIG=%s", kubeconfigPath)
+	restCfg, err := kubeconfig.GetNonInteractiveClientConfig(kubeconfigPath).ClientConfig()
+	m.Require().NoError(err, "Failed to clientFactory config")
+	m.clientFactory, err = client.NewSharedClientFactoryForConfig(restCfg)
+	m.Require().NoError(err, "Failed to create clientFactory Interface")
+
+	schemes.Register(v3.AddToScheme)
+	schemes.Register(provisioningv1.AddToScheme)
+
+	ns := &corev1.Namespace{
+		ObjectMeta: v1.ObjectMeta{
+			Name: testNamespace,
+		},
+	}
+	m.createObj(ns, schema.GroupVersionKind{
+		Group:   "",
+		Version: "v1",
+		Kind:    "Namespace",
+	})
+}
+
+func (m *IntegrationSuite) TearDownSuite() {
+	ns := &corev1.Namespace{
+		ObjectMeta: v1.ObjectMeta{
+			Name: testNamespace,
+		},
+	}
+	m.deleteObj(ns, schema.GroupVersionKind{
+		Group:   "",
+		Version: "v1",
+		Kind:    "Namespace",
+	})
+}
+
+// Object is generic object to wrap runtime and metav1.
+type Object interface {
+	v1.Object
+	runtime.Object
+}
+type endPointObjs[T Object] struct {
+	invalidCreate  func() T
+	validCreateObj T
+	newObj         func() T
+	invalidUpdate  func(obj T) T
+	validUpdate    func(obj T) T
+	invalidDelete  func() T
+	validDelete    func() T
+	gvk            schema.GroupVersionKind
+}
+
+func (m *IntegrationSuite) TestFeatureEndpoints() {
+	newObj := func() *v3.Feature { return &v3.Feature{} }
+	validCreateObj := &v3.Feature{
+		ObjectMeta: v1.ObjectMeta{
+			Name: "test-feature",
+		},
+		Spec: v3.FeatureSpec{Value: Ptr(false)},
+		Status: v3.FeatureStatus{
+			LockedValue: Ptr(false),
+			Description: "status description",
+		},
+	}
+	invalidUpdate := func(created *v3.Feature) *v3.Feature {
+		invalidUpdateObj := created.DeepCopy()
+		invalidUpdateObj.Spec.Value = Ptr(true)
+		return invalidUpdateObj
+	}
+	validUpdate := func(created *v3.Feature) *v3.Feature {
+		validUpdateObj := created.DeepCopy()
+		validUpdateObj.Status.Description = "Updated description"
+		return validUpdateObj
+	}
+	validDelete := func() *v3.Feature {
+		return validCreateObj
+	}
+	endPoints := &endPointObjs[*v3.Feature]{
+		invalidCreate:  nil,
+		newObj:         newObj,
+		validCreateObj: validCreateObj,
+		invalidUpdate:  invalidUpdate,
+		validUpdate:    validUpdate,
+		validDelete:    validDelete,
+	}
+	validateEndpoints(m.T(), endPoints, m.clientFactory)
+}
+
+func (m *IntegrationSuite) TestGlobalRole() {
```

Comment:
I would recommend splitting the test for various CRDs into separate files/packages of the integration test directory. The current structure sets a bad precedent for users who add integration tests in the future and may just expect to add more to a single file.